### PR TITLE
nvim_buf_set_text

### PR DESCRIPTION
**Pull request by chentoast**
This is a continuation of #12249, rebased on master.

Additional changes:

- fixed undo
- added some more tests
- removed the gravity changes - I'll probably follow up with that in a separate pr.
- updated all extmark functions to match the newer api.
- Added some documentation for extmark_splice, since I found that function to be kind of confusing at first. Let me know if I didn't get it right, or if the doc change is not wanted

### CODE REVIEW

**Comment by bfredl** [comment]
> Better handling of multibyte characters. It would be nice to have set_text work with virtual columns instead of actual columns.

Bytes is the standard way that neovim handles multibyte text. Are the are usecases you are thinking of that are not served by just converting the indexs prior to call if needed? (using whatever vcol or utf-16./32 distance api:s that already are available or added as separate API functions).

**Comment by chentau** [comment]
Yes, I think you’re right. It would probably be easier for plugins to process the bytes conversion using byteidx() or something like that.

**Comment by chentau** [comment]
One thing that I would like opinions on is how indexing beyond the end of the buffer should behave. Unlike buf_set_lines, the `end_row` of buf_set_text is inclusive. Therefore, setting `end_col` to -1 (index past end) should not be allowed, but this means that it would then be impossible to completely delete the last line - you can delete all of the text from the line, but the line itself will still be there.

I suppose the last point is not such a big problem, given that plugin authors can just use buf_set_lines instead, but it feels weird that buf_set_text can delete every line except for the last one.

@bfredl what do you think?

**Comment by bfredl** [comment]
@chentau hmm I am not sure if this is an actual problem, as you can still delete the NL _before_ the last line and thus effectively get rid of it. but I will need to do some interactive testing.

**Comment by chentau** [comment]
> you can still delete the NL _before_ the last line and thus effectively get rid of it. 

Tried this out and it does seem to work. Just for future reference, if we have the buffer

```
foo
bar
```

we can delete the last line via `nvim_buf_set_text(0, 0, 3, 1, 3, {})`, and insert a new line at the end via `nvim_buf_set_text(0, 1, 3, 1, 3, {'', 'hello'})`.

Assuming checks pass, I think this pr is ready for review.

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -487,6 +495,259 @@ void nvim_buf_set_lines(uint64_t channel_id,
...
+
+    // Same as with replacing, but we also need to free lines
+    xfree(lines[i]);
+    lines[i] = NULL;
+    extra++;
+  }
+
+  // Adjust marks. Invalidate any which lie in the
+  // changed range, and move any in the remainder of the buffer.
+  // Only adjust marks if we managed to switch to a window that holds
```
Comment (bfredl) [inline src/nvim/api/buffer.c:718]: this can be removed (we will always use a valid window, aucmd_win for hidden buffer).

**Comment by bfredl** [comment]
looks pretty solid. I will add a few tests for `on_bytes` as well.

**Comment by bfredl** [comment]
fixed some byte count issues, will test some more stuff after lunch :D

**Comment by bfredl** [comment]
merged as part of #11833, thanks!

**Comment by chentau** [comment]
Thanks for the help :)

**Pull request closed without merging**